### PR TITLE
Document time arrays

### PR DESCRIPTION
**Pull request by gautierronan**
This PR:

- updates and completes the tutorial on defining time arrays
- adds API documentation for TimeArray objects
- makes various stylistic changes required to document the class attributes and methods

2f9698afcc8a60bc140ff2a49253bea3fafbafd1 and efc5777cbdfa0f7790558ed369c239fbb9075abd are just copy-pasted. I recommand reviewing without these 2 commits.

Related to https://linear.app/dynamiqs/issue/DYN-147/improve-timearray-support.

### CODE REVIEW

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
This is neat. 😍 I've added a bunch of reformulations on top of your commits, happy to discuss anything you want to revert. 😊 I'll propagate the change to the tutorial page if you agree (and finish reviewing the tutorial page).

File: dynamiqs/time_array.py [hunk 1]
```diff
@@ -19,14 +19,14 @@
 def constant(array: ArrayLike) -> ConstantTimeArray:
     r"""Instantiate a constant time-array.
 
-    A constant time-array is defined by $A(t)=A_0$ for any time $t$, where $A_0$ is a
-    constant array.
+    A constant time-array is defined by $\hat O(t)= \hat O_0$ for any time $t$, where
+    $\hat O_0$ is a constant array.
```
[the remark this answers is not in the document]
Comment (gautierronan) [reply dynamiqs/time_array.py:23]: Agreed :)

**Comment by gautierronan** [comment]
LGTM, thanks for the corrections ! Feel free to merge.

**Comment by pierreguilmin** [comment]
A bunch of other fixes, I'll merge this now @gautierronan (40 commits 🤯), we can iterate on top. It's great, thanks a lot for the work 😍

**Pull request merged**